ipa-core/src/protocol/ipa_prf/shuffle/sharded.rs

```diff
@@ -233,15 +265,79 @@ impl<V: BooleanArray + FromRandom> Shuffleable for AdditiveShare<V> {
 }
 
 /// Trait for inputs to malicious shuffle.
-pub trait MaliciousShuffleable: Shuffleable {
+///
+/// Do not implement this trait directly. Implement `Shuffleable` and add an invocation
+/// of `impl_malicious_shuffle_share` for your `<T as Shuffleable>::Share` type, if it
+/// does not already exist.
+pub trait MaliciousShuffleable:
+    Shuffleable<Share = Self::MaliciousShare> + Clone + Default
+{
+    /// The `Shuffleable::Share` type, with additional bounds for malicious shuffle.
+    type MaliciousShare: ShuffleShare + MaliciousShuffleShare;
+
     /// A type that can hold `<Self as Shuffleable>::Share` along with a 32-bit MAC.
-    type ShareAndTag: BooleanArray + FromRandom;
+    type ShareAndTag: ShuffleShare + SharedValue;
+
+    /// The offset to the MAC in `ShareAndTag`.
+    const TAG_OFFSET: usize;
+
+    fn to_gf32bit(
+        &self,
+    ) -> Result<impl Iterator<Item = AdditiveShare<Gf32Bit>> + Send, crate::error::Error> {
+        let left_shares: Vec<Gf32Bit> = self.left().try_into()?;
+        let right_shares: Vec<Gf32Bit> = self.right().try_into()?;
+        Ok(left_shares
+            .into_iter()
+            .zip(right_shares)
+            .map(|(left, right)| ReplicatedSecretSharing::new(left, right)))
+    }
+}
+
+impl<T> MaliciousShuffleable for T
+where
+    T: Shuffleable + Clone + Default,
+    T::Share: MaliciousShuffleShare,
+{
+    type MaliciousShare = T::Share;
+    type ShareAndTag = <T::Share as MaliciousShuffleShare>::ShareAndTag;
+    const TAG_OFFSET: usize = <T::Share as MaliciousShuffleShare>::TAG_OFFSET;
 }
 
-impl MaliciousShuffleable for AdditiveShare<BA32> {
-    type ShareAndTag = BA64;
+/// Trait for a share of an input to malicious shuffle.
+///
+/// This trait should be implemented using the `impl_malicious_shuffleable!` macro,
+/// which will check the size of the `ShareAndTag` type and compute `TAG_OFFSET`
+/// automatically.
+pub trait MaliciousShuffleShare: TryInto<Vec<Gf32Bit>, Error = LengthError> {
+    type ShareAndTag: ShuffleShare + SharedValue;
```

Comment:
At some point it may be worth dropping `SharedValue` bound. It seems to me that we need it to be able to hold `AdditiveShare<Share+Tag>` to pass it down to semi-honest shuffle, but it is not really needed. We could just have our own container that is not an additive share and implement Shuffleable for it.

AdditiveShare carries luggage we don't need to bring into shuffle imo - we don't need complex arithmetic and many other bounds that it puts on us